### ISSUE TITLE
js/npm library?

### ISSUE THREAD

**Issue by nschloe** [penrose/penrose#608]
Not sure if I'm missing something here, but is there a JS library (on npm perhaps) that can be fed substance and style programs as strings and produces the figure in a given html element?

**Comment by maxkrieger** [comment]
Yup! Sorry for the lack of documentation here. Check out https://observablehq.com/@tophtucker/hello-penrose . The package is `@penrose/core@develop`.

**Comment by hypotext** [comment]
Added a link to wiki homepage: https://github.com/penrose/penrose/wiki

**Comment by nschloe** [comment]
Thanks for the links! However, from the tutorials etc. I can't quite see yet how to satisfy my use case. To rephrase, I would like to know if it's possible to define a Penrose diagram _in Javascript_ terms of three input strings or an input object, and get the resulting diagram as a Javascript object such that I can append it to an HTML element.

The most popular graphing/charting javascript libraries do this, e.g., [chartjs](https://www.chartjs.org/docs/latest/) via
```html
<canvas id="myChart" width="400" height="400"></canvas>
<script>
var ctx = document.getElementById('myChart').getContext('2d');
var myChart = new Chart(ctx, {
    type: 'bar',
    data: {
        labels: ['Red', 'Blue', 'Yellow', 'Green', 'Purple', 'Orange'],
        datasets: [{
            label: '# of Votes',
            data: [12, 19, 3, 5, 2, 3],
            backgroundColor: [
                'rgba(255, 99, 132, 0.2)',
                'rgba(54, 162, 235, 0.2)',
                'rgba(255, 206, 86, 0.2)',
                'rgba(75, 192, 192, 0.2)',
                'rgba(153, 102, 255, 0.2)',
                'rgba(255, 159, 64, 0.2)'
            ],
            borderColor: [
                'rgba(255, 99, 132, 1)',
                'rgba(54, 162, 235, 1)',
                'rgba(255, 206, 86, 1)',
                'rgba(75, 192, 192, 1)',
                'rgba(153, 102, 255, 1)',
                'rgba(255, 159, 64, 1)'
            ],
            borderWidth: 1
        }]
    },
    options: {
        scales: {
            y: {
                beginAtZero: true
            }
        }
    }
});
</script>
```

**Comment by wodeni** [comment]
> Thanks for the links! However, from the tutorials etc. I can't quite see yet how to satisfy my use case. To rephrase, I would like to know if it's possible to define a Penrose diagram _in Javascript_ terms of three input strings or an input object, and get the resulting diagram as a Javascript object such that I can append it to an HTML element.

Here's an example of using the UMD version of `@penrose/core` by @joshsunshine: https://www.cs.cmu.edu/~jssunshi/penrose/penrose-ide.html. 

The relevant bit in the source:
```html
<script type="text/javascript" src="./Penrose_files/penrose.min.js"></script>
<script type="text/javascript">
    function penrosify() {
        const elem = document.getElementById("root");
        const styArea = document.getElementById("styArea");
        const subArea = document.getElementById("subArea");
        const dslArea = document.getElementById("dslArea");
        const style = styArea.value;
        const substance = subArea.value;
        const domain = dslArea.value;
        elem.innerHTML = ""
        penrose.interactiveDiagram(domain, substance, style, elem);
    }
</script>
```

**Comment by nschloe** [comment]
Great! And noninteractive?

**Comment by wodeni** [comment]
> Great! And noninteractive?

There's an API function `diagram` with the same signature. See source here: https://github.com/penrose/penrose/blob/main/packages/core/src/index.ts#L71. Currently, the top-level API fucntions are documented [here](https://penrose.github.io/penrose/modules/_index_.html).

We are still working to keep the documentation and tutorials up-to-date. Please let us know if you have further feedback 😄 .

**Comment by nschloe** [comment]
I use @penrose/core 1.3.0 (`penrose.min.js`), and unfortunately,
```
penrose
```
is just an empty object for me. (It's defined though.) Any idea why that could be? In any case, distributing a non-minified version would be appreciated (see https://github.com/penrose/penrose/issues/620).

**Comment by nschloe** [comment]
I just found
```
Uncaught EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self'".
  at Function (<anonymous>)
    at Object.keywords (penrose.min.js:1)
    at penrose.min.js:1
    at penrose.min.js:1
    at penrose.min.js:1
```
so that's why the object is empty.

**Comment by wodeni** [comment]
> ```
> Uncaught EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self'".
>   at Function (<anonymous>)
>     at Object.keywords (penrose.min.js:1)
>     at penrose.min.js:1
>     at penrose.min.js:1
>     at penrose.min.js:1
> ```

Looks like this is coming from the `Function` object for the autodiff engine. @hypotext do you know any other alternatives to our current method?

**Comment by hypotext** [comment]
Not off the top of my head, it's pretty core to the way that we generate code for speed.

@nschloe what environment are you running the Penrose package in? Can you loosen the content security policy, e.g. here? https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/content_security_policy

It could also be due to a webpack setting @wodeni: https://stackoverflow.com/questions/48047150/chrome-extension-compiled-by-webpack-throws-unsafe-eval-error#49100966

**Comment by nschloe** [comment]
> what environment are you running the Penrose package in?

[A Chrome browser extension](https://github.com/nschloe/xdoc). The idea is to run Penrose sub/sty configs in GitHub READMEs, so there is user input involved. One would never want to allow execution of arbitrary user JS which is what the `eval()` ban is about. [It's perhaps possible to relax security](https://stackoverflow.com/a/26253927/353337), but I probably won't go down that route.

If I understand correctly, it's not user JS that penrose evaluates, but something else.

> It could also be due to a webpack setting 

Indeed. See also https://github.com/penrose/penrose/issues/620#issuecomment-875549181.

**Comment by samestep** [comment]
@nschloe Ah interesting, I didn't realize Chrome extensions had this restriction. As @k-qy mentioned, our code generation is key for keeping Penrose fast. However, it is possible that we may switch to generating WebAssembly instead of JavaScript. Do you happen to know if WebAssembly usage is allowed in Chrome extensions?

**Comment by nschloe** [comment]
> Do you happen to know if WebAssembly usage is allowed in Chrome extensions?

All user input that could potentially do harm on a computer of an extension user is prohibited. The only thing possible to to provide a string (or strings) to the extension that the extension then evaluates. (`eval()` and friends not allowed.)

**Comment by samestep** [comment]
@nschloe That makes sense; however, evaluating WebAssembly code has different security implications from evaluating JavaScript code. Judging from [this Stack Overflow answer](https://stackoverflow.com/a/49970231/5044950), it looks like it is possible to do inside of a Chrome extension. Actually, going off [Google's documentation](https://developer.chrome.com/docs/extensions/mv3/sandboxingEval/), it looks like it's even possible to use `eval` or `new Function` in some cases.

**Comment by nschloe** [comment]
> Actually, going off [Google's documentation](https://developer.chrome.com/docs/extensions/mv3/sandboxingEval/), it looks like it's even possible to use eval or new Function in some cases.

Quoting from the page:

> Published on Tuesday, September 18, 2012 • Updated on Thursday, May 22, 2014

and

> This page was migrated directly from the Manifest V2 documentation set. It has not yet been validated for compliance with Manifest V3.

The info is simply outdated.

**Comment by samestep** [comment]
Ah, that is unfortunate. But what I was asking about before was WebAssembly usage specifically; do you happen to know if the Stack Overflow thread I linked is also outdated? Or does that work today?

**Comment by nschloe** [comment]
> do you happen to know if the Stack Overflow thread I linked is also outdated? Or does that work today?

No idea.

**Comment by wodeni** [comment]
Hey @nschloe! Just wanted to give you an update on this issue. The autodiff engine has been rewritten to emit WebAssembly code. Although I'm not sure about the status of the requirements on Chrome extensions, here are some docs on using Penrose in [vanilla JS](https://penrose.cs.cmu.edu/docs/ref/vanilla-js) and [with bundlers](https://penrose.cs.cmu.edu/docs/ref/bundle).
